### ISSUE TITLE
Form submission data and playground output data do not match

### ISSUE THREAD

**Issue by Skaiir** [bpmn-io/form-js#273]
__Describe the Bug__

Currently what is displayed in the playground's second data window is the form state, but that does not correspond to the submission data that it claims to represent.

Examples I can see which are wrong:
- Forms do not submit disabled fields
- Forms only submit properties associated with a field

This isn't a simple case of just fixing this however, because 'current form state' is actually really useful, so I don't think we can just get rid of it. 

Perhaps a solution would be to make the submit button work in the playground, and open up the actual output in a second window? Or we could have three text panels but this becomes crowded. 

__Steps to Reproduce__

1. Open up the playground
2. Add a new randomly named property to `Form data (input) ` 
3. This gets populated to `Form data (submit)` despite not being configured in a field
4. Pressing submit in this case would not submit this data (see submission logic)

OR 

1. Open up the playground
2. Disable a field
3. Associated property isn't removed from `Form data (submit)`

![image](https://user-images.githubusercontent.com/9433996/189868580-06978ba7-09e9-41b5-842a-2164fb9e55ec.png)

__Expected Behavior__

Either this window should represent Form state, and be named as such, or it should represent Submission data, and should behave as such. Uncertain which is correct as of now.

**Comment by Skaiir** [comment]
@pinussilvestrus @nikku 

For visibility

**Comment by pinussilvestrus** [comment]
I had a second look at this issue, here are my observations

* In general, what we currently display in the `Form Data (Submit)` panel is the [_current form state_](https://github.com/bpmn-io/form-js/blob/master/packages/form-js-playground/src/components/PlaygroundRoot.js#L119), not the submit data.
* The label `Form Data (Submit)` is indeed misleading, as it does not show the submit data; so the data what you would get when you press on _Submit_. This is expected (cf. https://github.com/bpmn-io/form-js/issues/273#issuecomment-1245134467). 

Note that we only use this label in the core `form-js-playground`. We use a different in the `@camunda/form-playground` ([demo](https://camunda-form-playground.netlify.app/)).

![image](https://user-images.githubusercontent.com/9433996/189870035-844f06fb-f17f-4b9f-b010-aca428f39644.png)

**=> Proposal**: Let's update the label here as well to make it clear that it shows _Form Output Data_.


* > Forms do not submit disabled fields

I just checked that this is expected (cf. [`Form#submit`](https://github.com/bpmn-io/form-js/blob/develop/packages/form-js-viewer/src/Form.js#L179)). I added a screenshot to the issue description that shows the data is visible in the playground output data.

* > Forms only submit properties associated with a field

I can not verify this. If I add input data that is not used in any form field, it's also not showing up in the playground output data.

![image](https://user-images.githubusercontent.com/9433996/189870751-dc1fb09f-d967-480a-9c46-b87e874cccac.png)

I think we just removed this via https://github.com/bpmn-io/form-js/issues/266. We sanitize unknown values from the input data, so it is not visible on submit. However, as it is removed on import, and the playground form output panel relies on the current form state, it is also not showing up in the panel.

**=> Open for discussion:** Given we say the Form Output Data represents what is coming out of the form (e.g. process variables), shouldn't we also display unknown values there as well? I still think it should be erased on submit and the form itself should not bother about these value, but maybe we can display this data in the playground panel.

/cc @CatalinaMoisuc @christian-konrad @Skaiir @nikku

**Comment by nikku** [comment]
> I still think it should be erased on submit and the form itself should not bother about these value, but maybe we can display this data in the playground panel.

No!

What is the difference between _Output Data_ and _Submit Data_? Whatever our short term state is, mid term there should be none. The playground shall show what is being submitted (if the user did press submit). This includes only data that is recognized by the form, and excludes disabled fields; everything else is a bug. Why is this of such importance => Because it gives control over what is submitted to the form designer.

Submit behavior (and omission) is important to understand how form-js works. It shall not be ammended.

**Comment by pinussilvestrus** [comment]
> What is the difference between _Output Data_ and _Submit Data_? Whatever our short term state is, mid term there should be none.

To achieve that, we'd need to remove data from disabled form fields in the output panel, which is still visible currently (cf. screenshot above).

We need to clarify with @christian-konrad whether the state @nikku is mentioning (submission state) is the state we want to achieve.

**Comment by nikku** [comment]
If @christian-konrad envisions a different state I'd be interested how that state comes to be.

**Comment by christian-konrad** [comment]
@pinussilvestrus @nikku 

From my point of view, the output data panel should show exactly what will be submitted.
As a developer, I would not understand why there is a form state immediately before submit that is different than that on submit. I could only think of this if this would be a core part of the form lifecycle, including callbacks allowing me to alter things on submitting, and additional non-realtime validation before submit. I'd say we don't want this complexity—forms should be as reactive as possible (no hidden states appearing only after submit to server). I had really bad experiences with such behavior with a past form editor I was working on. Traceability is important—no suprises. @nikku happy to hear your take on his.

But, when we think about potential upcoming things like conditional rendering and formula fields, we may need to rethink some of the current behavior. For example, with formula fields, we may want disabled fields to be submitted, since the values of such fields are calculated and should not be changed by users. (At this point, I am not 100% sure if we should allow formula fields to be submitted; in case someone somehow injects malicious frontend code in custom tasklists, the calculation results may be wrong or malicious, and such things are better to be calculated using expressions in the output vars or in the job worker, but I leave it up to you to evaluate this once we work on this). There could also be hidden fields due to conditional rendering, and a developer wants the content of such fields still to be submitted (similar to hidden fields in element templates).

**Comment by Skaiir** [comment]
@christian-konrad @nikku @pinussilvestrus 

The way I see it, the problem we have here is simply that we have three important objects we're trying to fit into two boxes. I don't think we should debate choice or behavior here, all three have a purpose:

- Input Data / `Form Data (Input)` -- It simulates some external source initializing the form and acts as a 'clean slate' that we can always easily revert back to.
- Form State / `Form Data (Submit)` -- Lets a form builder play with the state to see how it affects the playground, this data will get messy and broken, which justifies the need for Input Data
- Output Data / `unrepresented` -- Displays what data the forms submits

Out of the three, the least important is the input data, whenever you edit it, it repopulates the form state directly so its true purpose functionally is simply to have an initial state to revert back to which does not get experimented upon like the form state will. If we didn't have this, we would have nothing to revert back to. 

I see three options, and all of them include keeping all three data structures in some form:

1. We have three panels, input, state, and output. I am not fond of that one because the UI is already cluttered enough, but it is easy to implement.
2. We have two panels for state and output, and input is tucked away as a configuration within the state panel, and we add a 'reset' to the state panel to revert back to the input state and clean up. I think this is the solution but it's also the hardest to implement properly, it will be a UX challenge. 
3. We have two panels for input and state, and output is requested via UI on the state panel, or on the form preview panel. This would also be relatively easy to implement, but not quite as nice as a good (2) solution.

I have a nice but relatively complex implementation in mind for (2), I can pitch it here but it's gonna need some sketches or a sync to get the point across. First I'd like to hear what your thoughts are on what I just said.

**Comment by nikku** [comment]
> Form State / Form Data (Submit) -- Lets a form builder play with the state to see how it affects the playground, this data will get messy and broken, which justifies the need for Input Data

This one does not exist. Why would I need it?

> Output Data / unrepresented -- Displays what data the forms submits

This one is what the output field is supposed to look like, albeit it is buggy.

**Comment by nikku** [comment]
> For example, with formula fields, we may want disabled fields to be submitted, since the values of such fields are calculated and should not be changed by users.

@christian-konrad I'd still argue that we don't ever want to submit disabled fields. This is not how the web-platform works.

**Comment by nikku** [comment]
Proposal: Let#s get together on a synchronous session to clarify the UX around forms / mental model + our direction for the playground.

Use KISS (keep it short and simple), our [design principles](https://github.com/bpmn-io/design-principles#no-surprises), expected web platform behavior as well as the embedding scenarios as a guideline.

**Comment by Skaiir** [comment]
@nikku 

> This one does not exist. Why would I need it?

I mean it exists in the sense that this is exactly what the `Form data (submit)` currently shows. It also exists under the hood in form-js's code. 

As to why we would need it, well to debug your form. As we start giving more powerful tool to form developers, such as conditional rendering, they will want to know why despite them entering X into Y component in the preview, Z is not hidden. For that they will want to inspect what them entering X did to the form state, check their conditional expression and see what they did wrong. If we solely rely on input data they are absolutely blind in that endeavor.

But yes, I agree on the sync, I think there's a lot to consider here.

**Comment by Skaiir** [comment]
And while I agree that the form output shows a lot of the data already, there will be cases where we have data within the form state upon which we'll want to conditionally render things but that are neither supplied by the inputs, nor outputted.

**Comment by nikku** [comment]
> As to why we would need it, well to debug your form. 

@Skaiir I try to take an extremely pragmatic attempt towards this. As of now we don't need such "internal state representation". If we neeeded "dynamic input debugging" is dumping the internal state the right way? I could imagine different options to help users understand why what component renders (or not). But let us tackle that once the use-case emerges. At least I don't see the necessity to do so now.

What I do know is that `input -> output` is of tremendous importance for our users, both standalone, but also in the embedding case (in Camunda Tasklist or the like).

**Comment by Skaiir** [comment]
@nikku Fair enough, we can live without it for now and tackle it when it becomes a necessity.

**Comment by christian-konrad** [comment]
>  I'd still argue that we don't ever want to submit disabled fields. This is not how the web-platform works.

@nikku could you please elaborate on what you mean with "This is not how the web-platform works"? Is it the html default that disabled fields are not submitted? How would we then give form developers the chance to submit additional data? What about rethinking the whole output mapping in general—probably this gets a little too far for now, but I think it makes sense to have "form output == task output". One idea: instead of submitting values of hidden fields, we would implicitly re-execute the expressions after form submit in the output mapping, without having the developer write the expression twice.  

The discussion is also a little tricky since we can not control the disabled state currently on runtime. We could possibly with conditional rendering. So far, hidden fields are not more than display fields, therefore it totally makes sense to not submit their content.

> My proposal would be: Let us get together on a synchronous session to clarify the UX around forms / mental model + our direction for the playground.

Would love to have such a chat.

@Skaiir 

> I mean it exists in the sense that this is exactly what the Form data (submit) currently shows. It also exists under the hood in form-js's code.

Let's question if this should be. IMHO, the current form state should be exactly what will be submitted. Then, it is easy for the developer to argue about what the form is doing, no surprises. But this would then also include to submit values of hidden and disabled fields, since they are part of that state. If we go for @nikku's proposal (only show form output and do not include hidden and disabled field values there), developers can't inspect intermediate state of conditional and formula fields.  But it's important to be able to inspect this state, since it has impact on what happens on next user interaction (think of the form state as a simple state machine with one idempotent operation "setValues(fieldValues, currentState) -> updatedState").

TLDR: Defining current form state == expected form output (and == user task output) helps future developers both to inspect their forms behavior and to not being suprised of the output.

**Comment by nikku** [comment]
The HTML way of submitting additional data is hidden inputs.

**Comment by nikku** [comment]
> Defining current form state == expected form output (and == user task output) helps future developers both to inspect their forms behavior and to not being suprised of the output.

Agreed, and also should match with the HTML form standard (let's discover differences!).

**Comment by Skaiir** [comment]
> But this would then also include to submit values of hidden and disabled fields, since they are part of that state.

And values lists, which we decided we didn't want to submit during dynamic data talks. Really any input data that we don't want in the output.

**Comment by pinussilvestrus** [comment]
Thanks for all your thoughts, folks ❤️ A couple of really important facts were stated, so I second your voices to get this one into a synchronous meeting, to define

* Which direction do we want to head _now_ (I suggest a pragmatic solution, as @nikku mentioned)?
* Which direction do we want to head _in the (near) future_ (when conditional rendering and friends arrive)?

I will schedule something will all of us.

**Comment by pinussilvestrus** [comment]
**Results of sync meeting with @nikku @christian-konrad @Skaiir** (17th October)

![image](https://user-images.githubusercontent.com/9433996/196179984-0b6336a0-a9c8-49b7-abc9-395ee34e78e4.png)

Miro board for more context: https://miro.com/app/board/uXjVPUN-mek=/

Next steps:
* We should discourage the use of disabled fields when possible, instead promote "ready-only" (cf. https://github.com/bpmn-io/form-js/issues/362)
* We should update the form output panel to _always display the submitted data_ => @pinussilvestrus will take this. This has a higher priority to include before we introduce the playground into the next Desktop Modeler release

/cc @CatalinaMoisuc who asked to stay in the loop.
